libpromises/rlist.c

```diff
@@ -1044,6 +1059,50 @@ Rlist *RlistFromSplitString(const char *string, char sep)
     return liststart;
 }
 
+/**
+ * Splits the given string into lines. On Windows, both \n and \r\n newlines are
+ * detected. Escaped newlines are respected/ignored too.
+ *
+ * @param detect_crlf whether to try to detect and respect "\r\n" line endings
+ * @return: an #Rlist where items are the individual lines **without** the
+ *          trailing newline character(s)
+ * @note: Free the result with RlistDestroy()
+ * @warning: This function doesn't work properly if @string uses "\r\n" newlines
+ *           and contains '\r' characters that are not part of any "\r\n"
+ *           sequence because it first splits @string on '\r'.
+ */
+Rlist *RlistFromStringSplitLines(const char *string, bool detect_crlf)
+{
+    if (string == NULL || string[0] == '\0')
+    {
+        return NULL;
+    }
+
+    if (!detect_crlf || (strstr(string, "\r\n") == NULL))
+    {
+        return RlistFromSplitString(string, '\n');
+    }
+
+    /* else we split on '\r' just like RlistFromSplitString() above, but
+     * strip leading '\n' in every chunk, thus effectively split on \r\n. See
+     * the warning in the function's documentation.*/
+    Rlist *liststart = NULL;
+
+    for (const char *sp = string; *sp != '\0';)
+    {
+        sp += SubStrnCopyChr(&liststart, sp, '\r', '\n');
```

Comment:
This looks a little scary, but I guess the only way `SubStrnCopyChr` can return `0` is when `*sp == '\0'`.

Reply:
That's a belief we share.